components/ingest-service/pipeline/publisher/msg_distributor.go

```diff
@@ -0,0 +1,86 @@
+package publisher
+
+import (
+	"time"
+
+	"github.com/chef/automate/components/ingest-service/backend"
+	"github.com/chef/automate/components/ingest-service/pipeline/message"
+)
+
+type ActivePipe struct {
+	in  chan message.ChefRun
+	out <-chan message.ChefRun
+}
+
+// BuildMsgDistributor a message distributor that sends all messages to the first child pipe until it is
+// full then start sending messages to the next child pipe in the line. When all the pipes are full the thread waits and
+// then checks to pipes again until it can place the message in a child pipe.
+func BuildMsgDistributor(childPipeBuilder message.ChefRunPipe, numProcessors, childPipeInboxSize int) message.ChefRunPipe {
+	return func(in <-chan message.ChefRun) <-chan message.ChefRun {
+		return fillOnePipeAtATime(in, childPipeBuilder, numProcessors, childPipeInboxSize)
+	}
+}
+
+func fillOnePipeAtATime(in <-chan message.ChefRun,
+	childPipeBuilder message.ChefRunPipe, numProcessors, childPipeInboxSize int) <-chan message.ChefRun {
+
+	out := make(chan message.ChefRun, 100)
+	pipes := buildActivePipes(childPipeBuilder, numProcessors, childPipeInboxSize)
+	mergeOutChannels(pipes, out)
+
+	go func() {
+		for msg := range in {
+			messageProcessed := false
+			for !messageProcessed {
+				for _, pipe := range pipes {
+					if !pipe.isFull() {
```

Comment:
have you considered using something like this so you don't need to check for space:
```
select {
case pipe.in <- msg:
   messageProcessed = true
default:
}
```

Reply:
this is fun: https://stackoverflow.com/questions/19992334/how-to-listen-to-n-channels-dynamic-select-statement

Reply:
Thank you. I will give that a try.

Reply:
The `select` worked great! Thanks Jay!

Reply:
I don't think you need the `default:` line

Reply:
@afiune You need the `default:` line to allow this line `pipe.in <- msg` to not wait for room in the channel. 

The below code is doing. If the channel is not full send the message and set `messageProcessed` to true. If the channel is full exit the select. 
```
case pipe.in <- msg:
   messageProcessed = true
default:
```

Reply:
Clever!!! ⭐️